src/nvim/screen.c

```diff
@@ -2317,93 +2317,115 @@ win_line (
-  int nextline_idx = 0;                 /* index in nextline[] where next line
-                                           starts */
-  int spell_attr = 0;                   /* attributes desired by spelling */
+  int nextline_idx = 0;                 /* index in nextline[] where next line starts */
   int word_end = 0;                     /* last byte with same spell_attr */
-  static linenr_T checked_lnum = 0;     /* line number for "checked_col" */
-  static int checked_col = 0;           /* column in "checked_lnum" up to which
-                                         * there are no spell errors */
-  static int cap_col = -1;              /* column to check for Cap word */
-  static linenr_T capcol_lnum = 0;      /* line number where "cap_col" used */
   int cur_checked_col = 0;              /* checked column for current line */
   int extra_check;                      /* has syntax or linebreak */
-  int multi_attr = 0;                   /* attributes desired by multibyte */
+
+  int c = 0;                            /* init for GCC */
   int mb_l = 1;                         /* multi-byte byte length */
   int mb_c = 0;                         /* decoded multi-byte character */
   int mb_utf8 = FALSE;                  /* screen char is UTF-8 char */
   int u8cc[MAX_MCO];                    /* composing UTF-8 chars */
+
+  #define SET_CHAR(ch) \
+    c = ch; \
+    mb_c = ch; \
+    mb_l = (*mb_char2len)(c); \
+    mb_utf8 = false; \
+    if (enc_utf8 && mb_l > 1) { \
+      c = 0xc0; \
+      mb_utf8 = true; \
+      u8cc[0] = 0; \
+    }
+
   int filler_lines;                     /* nr of filler lines to be drawn */
   int filler_todo;                      /* nr of filler lines still to do + 1 */
-  hlf_T diff_hlf = (hlf_T)0;            /* type of diff highlighting */
+  hlf_T diff_hlf = 0;                   /* type of diff highlighting */
+
   int change_start = MAXCOL;            /* first col of changed area */
   int change_end = -1;                  /* last col of changed area */
   colnr_T trailcol = MAXCOL;            /* start of trailing spaces */
-  int need_showbreak = FALSE;
-  int line_attr = 0;                    /* attribute for the whole line */
+  bool need_showbreak = false;
   matchitem_T *cur;                     /* points to the match list */
   match_T     *shl;                     /* points to search_hl or a match */
-  int shl_flag;                         /* flag to indicate whether search_hl
-                                           has been processed or not */
-  int prevcol_hl_flag;                  /* flag to indicate whether prevcol
-                                           equals startcol of search_hl or one
-                                           of the matches */
+  int shl_flag;                         /* flag to indicate whether search_hl has been processed or not */
+  int prevcol_hl_flag;                  /* flag to indicate whether prevcol equals startcol of search_hl or one of the matches */
```

Comment:
beware of the 80-column limit. If you reformat the comments, change them to `//` style.